### PR TITLE
Significantly decrease MemoryTracking drift

### LINKED ISSUE #15932: MemoryTracker wrong total: Memory limit (total) exceeded, but no real usage
Simple SELECT query leads to **Memory limit (total) exceeded** due to wrong reporting by MemoryTracker (no real usage detected in top)

I execute a simple query (`select a from b where c = 1`) in one thread sequentially via HTTP:
```
for i in `seq 1 500`; do curl "http://ch:8123/?query=SELECT%20a%20FROM%20%60db%60.b%20WHERE%20c%3D1%20format%20JSON"; done;
```
It leads to this error in 15 seconds
DB data ~200mb compressed.
It consumes ~4MB per request and accumulates up to the 3.6GB limit. I see a continuous growth in MemoryTracker, which, if you do not stop the requests (or switch to other request), will not be reset.
But there is no real growth in the `top`.
Result contains 0 rows if it matters
Version **20.9.2.20** stable, in 20.6 and 20.8 it was the same.

Changing config  max_server_memory_usage_to_ram_ratio is a dirty fix

### PR DESCRIPTION
**Pull request by azat**
Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Use total_memory_tracker when there is no other MemoryTracker object (this should significantly reduce the MemoryTracking drift)
- Fix parent memory tracker during query detaching
- Fix accounting for new/delete from different threads for VariableContext::Thread (supersedes #16122)

Detailed description / Documentation draft:
**After this PR the memory drift is ~0 after running 100 queries via http/tcp/tcp from one session. ([test](https://github.com/ClickHouse/ClickHouse/pull/16121/files#diff-70d939d0ce5a53be0dd9313cfeb0c805aaaa2fe8920e1f7e67cdc56bfd42c8ea))**

TL;DR;

<details>

To track memory, clickhouse creates memory tracker object for each
thread **explicitly**, but until it is not created the memory
allocations are not under account.
There should not be lot of allocations w/o memory tracker, since most of
the time it is created early enough, but even this maybe enough to
trigger some problems.

Plus sometimes it is not possible to create it, for example some 3d
party library does not allow to do this explicitly:
- for example before #15740 allocations from librdkafka threads,
- or even worse, poco threads, they don't have any routines to do this.
This won't be a problem for `MemoryTracking` metric if the deallocation
will be done from the same thread w/o memory tracker (or vise versa),
but this is not always true.

NOTE, that this will slow down per-thread allocations w/o memory
tracker, since before this patch there were no memory tracking for them
while now they will be accounted in total_memory_tracker, and for
total_memory_tracker max_untracked_memory is always reached.
But this should not be significant.

</details>

Changelog category (leave one):
- Not for changelog (changelog entry is not required)

<details>

HEAD:
- 7081b1047d50d403ae6ab4935f253107196464d9
- 990ad21b05254feae81341289e188d2d202b1ed1
- 43dd401ae5bd99ac954ad3d574c8f368c779b4f0 - worked, but problem in test
- 327d822b7f1b46b00d9f6167f389580a0ebbac0b - dump attempt to make the test pass

</details>

Fixes: #15932

### CODE REVIEW

File: src/Common/MemoryTracker.cpp [hunk 1]
```diff
@@ -6,13 +6,40 @@
...
+namespace
+{
+
+MemoryTracker * getMemoryTracker()
+{
+    if (auto * thread_memory_tracker = DB::CurrentThread::getMemoryTracker())
+        return thread_memory_tracker;
+
+    /// Do not use total_memory_tracker automatically in the main thread:
+    /// - there is not too much allocations in the main thread,
```
Comment (alexey-milovidov) [inline src/Common/MemoryTracker.cpp:26]: \- there is not too much allocations
\+ there are not too many allocations

**Comment by alexey-milovidov** [comment]
I like the idea, I have also tried to make something like that but then found it difficult.

**Comment by azat** [comment]
>I like the idea, I have also tried to make something like that but then found it difficult.

Ok, great (also it worth add a message when it is synced with RSS to see how it will differs later)

It still in draft due to gettid usage (that fact that performance does not shows the regression does not mean that there is no regression in this case, since most of the time total memory tracker is not used)

**Comment by alexey-milovidov** [comment]
You can also compare addresses of dummy thread_local variable (with the address of the same var from main).

**Comment by alexey-milovidov** [comment]
Also you can use atomic bool in constructor of thread-local variable and check it for non-zero.

File: src/Common/MemoryTracker.cpp (hunk 1 above)
Comment (azat) [reply src/Common/MemoryTracker.cpp:26]: Actually it will not be used there anyway, since it has it's own, so this comment will go away anyway

**Comment by azat** [comment]
Manual testing shows significant improvements, for http/tcp queries storm, I guess mosty due to memory accounting from the poco thread pools (require some test, will add later)

**Comment by azat** [comment]
Added a test that checks that after 100 queries the MemoryTracking does not changed (even to 1 byte) more then 6% (6 times), and it passed now (while before the metric is changed, and actually grows in some cases to ~4MB, after each query)!

File: tests/queries/0_stateless/01540_MemoryTracking.sh [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    changed="$(awk '{changed+=(prev && prev!=$0); prev=$0; } END {print changed}' 01540_MemoryTracking.memory.log)"
+
+    if [[ $changed -gt $changes_allowed ]]; then
+        echo "Test $test_func failed. MemoryTracking changed too frequently: $changed (allowed $changes_allowed)" >&2
+    fi
+}
+
+# wait until the server will be initialized
+# (since it may do some costly allocations)
+sleep 1m
```
Comment (akuzm) [inline tests/queries/0_stateless/01540_MemoryTracking.sh:83]: Looks like this is the first line of the script that is ran, so what is it waiting for?

Comment (azat) [reply tests/queries/0_stateless/01540_MemoryTracking.sh:83]: Yes, this is just an attempt in background, while I was digging into the issue (that was `log_queries`), this patch will be removed shortly.

**Comment by azat** [comment]
So the problem was only in the test

Due to metric_log it has some difference, of course I could rewrite the test to allow some small delta, ~20K is enough, but I want ensure/show that after some actions the MemoryTracking does not changes at all.
So the stateless test was replaced with integration test (basically it is needed only to run instance of the server w/o `system.*_log`)

**Comment by azat** [comment]
So now looks OK (although there is minimal conflict in the fasttest skip list, it is pretty trivial so I will not solve it by myself to avoid trigerring CI)

File: src/Common/MemoryTracker.cpp [hunk 3]
```diff
@@ -192,14 +210,15 @@ void MemoryTracker::free(Int64 size)
         DB::TraceCollector::collect(DB::TraceType::MemorySample, StackTrace(), -size);
     }
 
+    Int64 accounted_size = size;
```
Comment (alexey-milovidov) [inline src/Common/MemoryTracker.cpp:213]: Don't understand this change.

File: src/Interpreters/ThreadStatusExt.cpp [hunk 4]
```diff
@@ -301,7 +301,7 @@ void ThreadStatus::detachQuery(bool exit_if_already_detached, bool thread_exits)
     memory_tracker.reset();
 
     /// Must reset pointer to thread_group's memory_tracker, because it will be destroyed two lines below.
-    memory_tracker.setParent(nullptr);
+    memory_tracker.setParent(thread_group->memory_tracker.getParent());
```
Comment (alexey-milovidov) [inline src/Interpreters/ThreadStatusExt.cpp:304]: Commend becomes obsolete.

File: src/Common/MemoryTracker.cpp (hunk 3 above)
Comment (alexey-milovidov) [reply src/Common/MemoryTracker.cpp:213]: Ok.

**Review by alexey-milovidov — APPROVED** [review]
